From: Jeff Mahoney <jeffm@suse.com>
Subject: [PATCH 12/15] ocfs2: only disconnect tcp connections when the last reference is dropped

 Since o2net wants all events from any resource, by default it will drop the
 TCP connection when a single resource fails.

 This patch waits until all listeners have been dropped to drop the
 connection.

 fs/ocfs2/cluster/nodemanager.h |    1 +
 fs/ocfs2/cluster/tcp.c         |    7 +++++--
 2 files changed, 6 insertions(+), 2 deletions(-)

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

--- a/fs/ocfs2/cluster/nodemanager.h	2007-07-08 19:32:17.000000000 -0400
+++ b/fs/ocfs2/cluster/nodemanager.h	2007-08-27 14:03:23.000000000 -0400
@@ -50,6 +50,7 @@ struct o2nm_node {
 	int			nd_local;
 
 	unsigned long		nd_set_attributes;
+	atomic_t		nd_count;
 };
 
 struct o2nm_cluster {
--- a/fs/ocfs2/cluster/tcp.c	2007-08-27 14:01:22.000000000 -0400
+++ b/fs/ocfs2/cluster/tcp.c	2007-08-27 14:01:22.000000000 -0400
@@ -1590,8 +1590,10 @@ void o2net_disconnect_node(struct o2nm_n
 static void o2net_hb_node_down_cb(struct o2nm_node *node, int node_num,
 				  void *data)
 {
-	if (node_num != o2nm_this_node())
-		o2net_disconnect_node(node);
+	if (node_num != o2nm_this_node()) {
+		if (atomic_dec_and_test(&node->nd_count))
+			o2net_disconnect_node(node);
+	}
 
 	BUG_ON(atomic_read(&o2net_connected_peers) < 0);
 }
@@ -1606,6 +1608,7 @@ static void o2net_hb_node_up_cb(struct o
 		(msecs_to_jiffies(o2net_reconnect_delay(node)) + 1);
 
 	if (node_num != o2nm_this_node()) {
+		atomic_inc(&node->nd_count);
 		/* heartbeat doesn't work unless a local node number is
 		 * configured and doing so brings up the o2net_wq, so we can
 		 * use it.. */
